cli/stat.go

```diff
@@ -0,0 +1,170 @@
+package cli
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/shirou/gopsutil/cpu"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/cli/clibase"
+	"github.com/coder/coder/cli/cliui"
+)
+
+type statCmd struct {
+	*RootCmd
+	watch time.Duration
+}
+
+func (r *RootCmd) stat() *clibase.Cmd {
+	c := &clibase.Cmd{
+		Use:   "stat <type> [flags...]",
+		Short: "Display local system resource usage statistics",
+		Long:  "stat calls can be used as the script for agent metadata blocks.",
+	}
+	sc := statCmd{RootCmd: r}
+	c.Options.Add(
+		clibase.Option{
+			Flag:          "watch",
+			FlagShorthand: "w",
+			Description:   "Continuously display the statistic on the given interval.",
+			Value:         clibase.DurationOf(&sc.watch),
+		},
+	)
+	c.AddSubcommands(
+		sc.cpu(),
+	)
+	sc.setWatchLoops(c)
+	return c
+}
+
+func (sc *statCmd) setWatchLoops(c *clibase.Cmd) {
+	for _, cmd := range c.Children {
+		innerHandler := cmd.Handler
+		cmd.Handler = func(inv *clibase.Invocation) error {
+			if sc.watch == 0 {
+				return innerHandler(inv)
+			}
+
+			ticker := time.NewTicker(sc.watch)
+			defer ticker.Stop()
+
+			for range ticker.C {
+				if err := innerHandler(inv); err != nil {
+					_, _ = fmt.Fprintf(inv.Stderr, "error: %v", err)
+				}
+			}
+			panic("unreachable")
+		}
+	}
+}
+
+func cpuUsageFromCgroup(interval time.Duration) (float64, error) {
+	cgroup, err := os.OpenFile("/proc/self/cgroup", os.O_RDONLY, 0)
+	if err != nil {
+		return 0, err
+	}
+	defer cgroup.Close()
+	sc := bufio.NewScanner(cgroup)
+
+	var groupDir string
+	for sc.Scan() {
+		fields := strings.Split(sc.Text(), ":")
+		if len(fields) != 3 {
+			continue
+		}
+		if fields[1] != "cpu,cpuacct" {
+			continue
+		}
+		groupDir = fields[2]
+		break
+	}
+
+	if groupDir == "" {
+		return 0, xerrors.New("no cpu cgroup found")
+	}
+
+	cpuAcct := func() (int64, error) {
+		path := fmt.Sprintf("/sys/fs/cgroup/cpu,cpuacct/%s/cpuacct.usage", groupDir)
```

Comment:
@bpmct go pkg doesn't handle this on its own, and we need to specify the path of the cgroup, which is kernel/release dependent.